lib/recursive-methods.rb

```diff
@@ -1,49 +1,98 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n is equal to the input value
+# Space complexity: O(n) - where n is equal to the input value
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, "input must be a number greater than or equal to 0."
+  elsif n == 1 || n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n is equal to the length of the input string
+# not confident on this space complexity
+# Space complexity: O(n ^ 2) - where n is equal to the length of the input string ^ 2
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0 || s.length == 1
+    return s
+  else
+    return s[-1] + (reverse(s[0..-2]))
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n) - where n is the length of the string / 2
+# Space complexity: O(n) - where n is the length of the string / 2
+def reverse_inplace(s, i = 0, j = s.length - 1)
+  if i >= j
+    return s
+  else
+    temp = s[i]
+    s[i] = s[j]
+    s[j] = temp
+
+    return reverse_inplace(s, i += 1, j -= 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n is equal to the input value
+# Space complexity: O(n) - where n is equal to the input value
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return n
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - where n equal the length of the string / 2
```

Comment:
Because `s[1..-2]` creates a new string of length n-2, this is an O(n^2) for both space and time.